### PR TITLE
[RAG] Propagating of n_docs as parameter to all RagModel's related functions

### LINKED ISSUE #7874: [Rag] extend_enc_output fails when number of retrieved documents not equal to RagConfig.n_docs
## Environment info

     
- `transformers` version: 3.3.1
- Platform: Linux-4.19.112+-x86_64-with-Ubuntu-18.04-bionic
- Python version: 3.6.9
- PyTorch version (GPU?): 1.6.0+cu101 (False)
- Tensorflow version (GPU?): 2.3.0 (False)
- Using GPU in script?: No
- Using distributed or parallel set-up in script?: No

### Who can help

@patrickvonplaten

## Information

Model I am using (Bert, XLNet ...): Rag

The problem arises when using:
* [X] my own modified scripts: (give details below)

The tasks I am working on is:
* [X] an official GLUE/SQUaD task: `dummy_dataset`

## To reproduce

1. Use caling retriever seperately example and but modify number of retrieved document parameter `n_docs` as follows -
```
docs_dict = retriever(input_ids.numpy(), question_hidden_states.detach().numpy(), return_tensors="pt", n_docs=2)
```

Code snippet -
```
tokenizer = RagTokenizer.from_pretrained("facebook/rag-token-nq")
retriever = RagRetriever.from_pretrained("facebook/rag-token-nq", index_name="exact", use_dummy_dataset=True)
model = RagTokenForGeneration.from_pretrained("facebook/rag-token-nq", use_dummy_dataset=True)

input_dict = tokenizer.prepare_seq2seq_batch("What is capital of France?", return_tensors="pt")
input_ids = input_dict["input_ids"]

# Caling retriever seperately
question_hidden_states = model.question_encoder(input_ids)[0]

# Default value of RagConfig.n_docs is 5 hence change n_docs to different than default value
docs_dict = retriever(input_ids.numpy(), question_hidden_states.detach().numpy(), return_tensors="pt", n_docs=2)
doc_scores = torch.bmm(question_hidden_states.unsqueeze(1), docs_dict["retrieved_doc_embeds"].float().transpose(1, 2)).squeeze(1)

outputs = model.generate(context_input_ids=docs_dict["context_input_ids"], context_attention_mask=docs_dict["context_attention_mask"], doc_scores=doc_scores)
generated_string = tokenizer.batch_decode(outputs, skip_special_tokens=True)
print(generated_string)
```
Stacktrace -
```
/usr/local/lib/python3.6/dist-packages/transformers/modeling_rag.py in generate(self, input_ids, attention_mask, context_input_ids, context_attention_mask, doc_scores, max_length, min_length, early_stopping, use_cache, num_beams, bos_token_id, pad_token_id, eos_token_id, length_penalty, no_repeat_ngram_size, bad_words_ids, num_return_sequences, decoder_start_token_id, **kwargs)
   1353 
   1354         # correctly extend last_hidden_state and attention mask
-> 1355         context_attention_mask = extend_enc_output(context_attention_mask, num_beams=num_beams)
   1356         encoder_outputs["last_hidden_state"] = extend_enc_output(last_hidden_state, num_beams=num_beams)
   1357 

/usr/local/lib/python3.6/dist-packages/transformers/modeling_rag.py in extend_enc_output(tensor, num_beams)
   1346         def extend_enc_output(tensor, num_beams=None):
   1347             # split into `batch_size`, `num_beams`, `num_docs`
-> 1348             tensor = tensor[None, None, :].reshape((batch_size, 1, self.config.n_docs) + tensor.shape[1:])
   1349             # repeat same last hidden states over `num_beams` dimension
   1350             tensor = tensor.expand((batch_size, num_beams, self.config.n_docs) + tensor.shape[3:])

RuntimeError: shape '[0, 1, 5, 300]' is invalid for input of size 600
```
## Expected behavior

I think `extend_enc_output` should use tensor dimension instead of  `self.config.n_docs` as follows -
```
tensor = tensor[None, None, :].reshape((batch_size, 1, tensor.shape[0:]) + tensor.shape[1:])
```
Let me know if proposed fix is fine, I can raise PR.

### PR DESCRIPTION
**Pull request by lalitpagaria**
# What does this PR do?

Fixes #7874

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dimiss the other checks if that's the case).
- [X] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests), 
      Pull Request section?
- [X] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to the it if that's the case.
- [X] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [X] Did you write any new necessary tests? 

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors which may be interested in your PR.

@patrickvonplaten

### CODE REVIEW

File: src/transformers/modeling_rag.py [hunk 1]
```diff
@@ -458,6 +458,8 @@ def from_pretrained_question_encoder_generator(
         output_retrieved(:obj:`bool`, `optional`):
             Whether or not to return the :obj:`retrieved_doc_embeds`, :obj:`retrieved_doc_ids`,
             :obj:`context_input_ids` and :obj:`context_attention_mask`. See returned tensors for more detail.
+        n_docs (:obj:`int`, `optional`, defaults to `config.n_docs`)
+            Number of documents are retrieved or to be retrieved from the retriever.
```
Comment (patrickvonplaten) [inline src/transformers/modeling_rag.py:462]:
```suggestion
            Number of documents to retrieve and/or number of documents for which to generate an answer.
```

**Review by patrickvonplaten — APPROVED** [review]
Besides a small suggestion for the docstrings this PR looks great! Thanks a lot @lalitpagaria !

**Comment by patrickvonplaten** [comment]
@lhoestq would be great if you can review as well

**Comment by lalitpagaria** [comment]
@patrickvonplaten Thanks for the review.

while working on this PR I found that in `RagTokenForGeneration` we are computing `batch_size` as follows -
```
batch_size = context_input_ids.shape[0] // n_docs
```
So still issue can come when `((context_input_ids.shape[0] % n_docs) != 0)`, but I can't think of solution to address this.

File: src/transformers/modeling_rag.py [hunk 2]
```diff
@@ -1274,13 +1298,16 @@ def generate(
                 function, where we set ``num_return_sequences`` to :obj:`num_beams`.
             decoder_start_token_id (:obj:`int`, `optional`):
                 If an encoder-decoder model starts decoding with a different token than `bos`, the id of that token.
+            n_docs (:obj:`int`, `optional` defaults to self.config.n_docs)
+                Number of documents to retrieved
```
Comment (lhoestq) [inline src/transformers/modeling_rag.py:1302]:
```suggestion
                Number of documents to retrieve
```

**Review by lhoestq — APPROVED** [review]
Thanks, looks good to me :)

**Comment by lhoestq** [comment]
> @patrickvonplaten Thanks for the review.
> 
> while working on this PR I found that in `RagTokenForGeneration` we are computing `batch_size` as follows -
> 
> ```
> batch_size = context_input_ids.shape[0] // n_docs
> ```
> 
> So still issue can come when `((context_input_ids.shape[0] % n_docs) != 0)`, but I can't think of solution to address this.

`context_input_ids` is always supposed to have a size of `n_docs` times the number of input questions

File: src/transformers/modeling_rag.py (hunk 2 above)
Comment (lalitpagaria) [reply src/transformers/modeling_rag.py]: Thanks done

**Comment by lalitpagaria** [comment]
> > @patrickvonplaten Thanks for the review.
> > while working on this PR I found that in `RagTokenForGeneration` we are computing `batch_size` as follows -
> > ```
> > batch_size = context_input_ids.shape[0] // n_docs
> > ```
> > 
> > 
> > So still issue can come when `((context_input_ids.shape[0] % n_docs) != 0)`, but I can't think of solution to address this.
> 
> `context_input_ids` is always supposed to have a size of `n_docs` times the number of input questions

It would be better if we mention it explicitly by assert. WDYT? 
In one of my test case I used `n_docs=3` for retriever and `n_docs=2` for generator and it failed

**Comment by lhoestq** [comment]
> It would be better if we mention it explicitly by assert. WDYT?
> In one of my test case I used `n_docs=3` for retriever and `n_docs=2` for generator and it failed

Yes indeed. Also if `((context_input_ids.shape[0] % n_docs) != 0)` then we should raise an error otherwise some retrieved documents will be ignored for generation.

**Comment by patrickvonplaten** [comment]
Yes @lalitpagaria - it would be nice if you can add an asserte statement verifying that `n_docs` is correctly set. `n_docs` should be the same for both retriever and generator.

**Comment by lalitpagaria** [comment]
@patrickvonplaten @lhoestq Added assert at two places please verify, along with supporting unit test. Pardon my naming convention for test function, and please suggest proper name :) 

> n_docs should be the same for both retriever and generator.

This can't be check if `generator` does not know about `retriever` hence using this `((context_input_ids.shape[0] % n_docs) != 0)`

File: tests/test_modeling_rag.py [hunk 3]
```diff
@@ -186,10 +186,10 @@ def tearDown(self):
...
         dataset = Dataset.from_dict(
             {
-                "id": ["0", "1"],
-                "text": ["foo", "bar"],
-                "title": ["Foo", "Bar"],
-                "embeddings": [np.ones(self.retrieval_vector_size), 2 * np.ones(self.retrieval_vector_size)],
+                "id": ["0", "1", "3"],
+                "text": ["foo", "bar", "qux"],
+                "title": ["Foo", "Bar", "Qux"],
+                "embeddings": [np.ones(self.retrieval_vector_size), 3 * np.ones(self.retrieval_vector_size)],
```
Comment (lhoestq) [inline tests/test_modeling_rag.py:192]: This list should have three elements instead of 2

Comment (lalitpagaria) [reply tests/test_modeling_rag.py]: Thanks fixed this

**Comment by lalitpagaria** [comment]
@patrickvonplaten and @lhoestq Thank for review. I liked the test coverage of this project. Initially I struggled but letter all worked nicely. You can merge when you want.

File: src/transformers/modeling_rag.py [hunk 4]
```diff
@@ -458,6 +458,8 @@ def from_pretrained_question_encoder_generator(
         output_retrieved(:obj:`bool`, `optional`):
             Whether or not to return the :obj:`retrieved_doc_embeds`, :obj:`retrieved_doc_ids`,
             :obj:`context_input_ids` and :obj:`context_attention_mask`. See returned tensors for more detail.
+        n_docs (:obj:`int`, `optional`, defaults to `config.n_docs`)
```
Comment (patrickvonplaten) [inline src/transformers/modeling_rag.py:461]:
```suggestion
        n_docs (:obj:`int`, `optional`, defaults to :obj:`config.n_docs`)
```

File: src/transformers/modeling_rag.py [hunk 5]
```diff
@@ -1329,8 +1360,12 @@ def generate(
                 1
             )
 
+        assert (
+            context_input_ids.shape[0] % n_docs
+        ) == 0, "Make sure `context_input_ids` is always have a size of `n_docs` times the number of input questions."
```
Comment (patrickvonplaten) [inline src/transformers/modeling_rag.py:1365]:
```suggestion
        ) == 0, f" The first dimension of `context_input_ids` should be a multiple of `n_docs`={n_docs}, but is {context_input_ids.shape[0]}."
```

File: src/transformers/modeling_rag.py [hunk 6]
```diff
@@ -600,12 +604,16 @@ def forward(
             doc_scores is not None
         ), "Make sure that `doc_scores` are passed when passing `encoder_outputs` to the forward function."
 
+        assert (
+            doc_scores.shape[1] % n_docs
+        ) == 0, "Make sure `doc_scores.shape[1]` is always have a size of `n_docs`."
```
Comment (patrickvonplaten) [inline src/transformers/modeling_rag.py:609]:
```suggestion
        ) == 0, f" The first dimension of `context_input_ids` should be a multiple of `n_docs`={n_docs}, but is {context_input_ids.shape[0]}."
```

File: src/transformers/modeling_rag.py [hunk 7]
```diff
@@ -847,6 +860,8 @@ def generate(
                 function, where we set ``num_return_sequences`` to :obj:`num_beams`.
             num_beams (:obj:`int`, `optional`, defaults to 1):
                 Number of beams for beam search. 1 means no beam search.
+            n_docs (:obj:`int`, `optional`, defaults to `config.n_docs`)
```
Comment (patrickvonplaten) [inline src/transformers/modeling_rag.py:863]:
```suggestion
            n_docs (:obj:`int`, `optional`, defaults to :obj:`config.n_docs`)
```

File: src/transformers/modeling_rag.py [hunk 8]
```diff
@@ -1274,13 +1302,16 @@ def generate(
                 function, where we set ``num_return_sequences`` to :obj:`num_beams`.
             decoder_start_token_id (:obj:`int`, `optional`):
                 If an encoder-decoder model starts decoding with a different token than `bos`, the id of that token.
+            n_docs (:obj:`int`, `optional`, defaults to `config.n_docs`)
```
Comment (patrickvonplaten) [inline src/transformers/modeling_rag.py:1305]:
```suggestion
            n_docs (:obj:`int`, `optional`, defaults to :obj:`config.n_docs`)
```

**Comment by patrickvonplaten** [comment]
Slow tests pass => ready to merge

**Comment by patrickvonplaten** [comment]
Good job @lalitpagaria !

File: src/transformers/modeling_rag.py (hunk 6 above)
Comment (lalitpagaria) [reply src/transformers/modeling_rag.py]: @patrickvonplaten Please review your this change change, while debugging I have seen `context_input_ids == None` at this point. Specially while using `RagSequenceForGeneration`  without `retriever`. Hence it should be -
```
f" The first dimension 1 of `doc_scores` should be a multiple of `n_docs`={n_docs}, but is {doc_scores.shape[1]}."
```

**Pull request merged**